src/nn/wasm/src/common/operations/Conv2D.cpp

```diff
@@ -0,0 +1,103 @@
+                                                                                \
+    im2colDim.strides[0] = 1;                                                   \
+    for (int i=1; i<4; i++) {                                                   \
+        im2colDim.strides[i] = im2colDim.strides[i-1] * im2colDim.sizes[i-1];   \
+    }                                                                           \
+                                                                                \
+    Type* im2colData = nullptr;                                                 \
+    uint64_t im2colByteSize = sizeof(Type);                                     \
+    std::unique_ptr<Type[]> im2colGuard;                                        \
+    for (int i=0; i<4; i++) {                                                   \
+        im2colByteSize *= im2colDim.sizes[i];                                   \
+    }                                                                           \
+    /* http://b/77982879, tflite::optimized_ops::Conv uses int for offsets */   \
+    if (im2colByteSize >= 0x7fffffff)  {                                        \
+        LOG(ERROR) << "Conv size is too large, not enough memory";              \
+        return false;                                                           \
+    }                                                                           \
+    if (im2colByteSize <= kStaticBufferSize) {                                  \
+        im2colData = reinterpret_cast<Type *>(static_scratch_buffer);           \
+    } else {                                                                    \
+        im2colData = new (std::nothrow) Type[im2colByteSize / sizeof(Type)];    \
+        if (im2colData == nullptr) {                                            \
+            LOG(ERROR) << "Conv size is too large, not enough memory";          \
+            return false;                                                       \
+        }                                                                       \
+        im2colGuard.reset(im2colData);                                          \
+    }
+
+bool convFloat32(const float* inputData, const Shape& inputShape,
+                 const float* filterData, const Shape& filterShape,
+                 const float* biasData, const Shape& biasShape,
+                 int32_t padding_left, int32_t padding_right,
+                 int32_t padding_top, int32_t padding_bottom,
+                 int32_t stride_width, int32_t stride_height,
+                 int32_t activation,
+                 float* outputData, const Shape& outputShape) {
+
+    ANDROID_NN_CONV_PARAMETERS(float)
+
+    float output_activation_min, output_activation_max;
+    CalculateActivationRangeFloat(activation, &output_activation_min,
+                                  &output_activation_max);
+    
+    int32_t dilationWidthFactor = 1, dilationHeightFactor = 1;
+
+    tflite::optimized_ops::Conv(
```

Comment:
Can we expose the `tflite::optimized_ops::Conv` and `tflite::Dims` directly? So the js code can call it without this Android NN wrapper code.

Reply:
If it works, we can apply it to other ops and simplify our binding code.

Reply:
I don't know if I understand it correctly. Do you mean calling TFLite code directly in the binding code?

If that's the case, we may still need to move these NN code to the bindings in dealing these parameters. Or should we make these Android wrappers`inline`?

Reply:
No. I mean call tflite functions directly in JS. For example, call `tflite::optimized_ops::Conv` 
 [here](https://github.com/intel/webml-polyfill/blob/master/src/nn/wasm/PreparedModel.js#L196). No need NN wrapper code anymore.

Reply:
Ok, I see what you mean. We'll keep working on it.

Reply:
Thanks, I will first have a try with Conv2D!